llvm/include/llvm/Analysis/KernelInfo.h

```diff
@@ -0,0 +1,121 @@
+//=- KernelInfo.h - Kernel Analysis -------------------------------*- C++ -*-=//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the KernelInfo, KernelInfoAnalysis, and KernelInfoPrinter
+// classes used to extract function properties from a GPU kernel.
+//
+// See llvm/docs/KernelInfo.rst.
+// ===---------------------------------------------------------------------===//
+
+#ifndef LLVM_ANALYSIS_KERNELINFO_H
+#define LLVM_ANALYSIS_KERNELINFO_H
+
+#include "llvm/Analysis/OptimizationRemarkEmitter.h"
+#include "llvm/Analysis/TargetTransformInfo.h"
+
+namespace llvm {
+class DominatorTree;
+class Function;
+
+/// Data structure holding function info for kernels.
+class KernelInfo {
```

Comment:
Is there any real use of this beyond running the remark emitter? Sink this into the implementation file?

Reply:
I'm curious as well. If there is no, I really don't think it should be in a target independent way that merges attributes from all targets together.

Reply:
wrt. other uses: Likely, in the future. For now, we can sink it into the impl file.
wrt. merging attributes: There is really nothing to be won by splitting this into 3-4 files. It just adds complexity.

Reply:
It doesn't have to be multiple files. The variadic argument pass has target dependent stuff in the same file as well. 

What I expect is a hierarchical modular data struct. Each target just minds their own business: add their own attributes in their own descriptor, implement their own analysis and print method, etc. For now it "seems" okay to have all of them in the same data structure and use if-else to print, because technically we only have two offloading targets, but this data structure will keep growing when we have more (looking at Intel), and for any of the target, most of the parts are useless.

For the future use, we can just have the analysis results at one place and return a reference.

Reply:
> For now, we can sink it into the impl file.

In that case, does anyone see value in keeping the KernelInfoAnalysis class for now?  KernelInfoPrinter can just call KernelInfo directly.

Reply:
> > For now, we can sink it into the impl file.
> 
> In that case, does anyone see value in keeping the KernelInfoAnalysis class for now? KernelInfoPrinter can just call KernelInfo directly.

I've dropped KernelInfoAnalysis for now.

Reply:
@shiltian The [commit I just added](https://github.com/llvm/llvm-project/pull/102944/commits/62d494d9a9f13e5b58d71f083a6cb9f67f19579b) to address another concern probably also address your concern about target modularity.

Reply:
Resolving this issue based on my previous comment here.  Please reopen if necessary.